crates/autopilot/src/infra/persistence/mod.rs

```diff
@@ -300,10 +283,97 @@ impl Persistence {
+    ///
+    /// It is expected for a solution to exist, so missing data is considered an
+    /// error.
+    ///
+    /// Returns error for old non-colocated auctions.
+    pub async fn get_winning_solution(
+        &self,
+        auction_id: domain::auction::Id,
+    ) -> Result<domain::competition::Solution, error::Solution> {
+        let _timer = Metrics::get()
+            .database_queries
+            .with_label_values(&["get_competition_winner"])
+            .start_timer();
+
+        let mut ex = self
+            .postgres
+            .pool
+            .begin()
+            .await
+            .map_err(error::Solution::BadCommunication)?;
+
+        let competition = database::settlement_scores::fetch(&mut ex, auction_id)
+            .await
+            .map_err(error::Solution::BadCommunication)?
+            .ok_or(error::Solution::NotFound)?;
+
+        let winner = H160(competition.winner.0).into();
+        let score = competition::Score::new(
+            big_decimal_to_u256(&competition.winning_score)
+                .ok_or(error::Solution::InvalidScore(anyhow::anyhow!(
+                    "database score"
+                )))?
+                .into(),
+        )
+        .map_err(|_| error::Solution::InvalidScore(anyhow::anyhow!("zero score")))?;
+
+        let solution = {
+            // TODO: stabilize the solver competition table to get promised solution.
+            let solver_competition = database::solver_competition::load_by_id(&mut ex, auction_id)
+                .await
+                .map_err(error::Solution::BadCommunication)?
+                .ok_or(error::Solution::NotFound)?;
+            let competition: model::solver_competition::SolverCompetitionDB =
+                serde_json::from_value(solver_competition.json)
+                    .context("deserialize SolverCompetitionDB")
+                    .map_err(error::Solution::InvalidSolverCompetition)?;
+            let winning_solution = competition
+                .solutions
+                .last()
+                .ok_or(error::Solution::NotFound)?;
+            let mut orders = HashMap::new();
+            for order in winning_solution.orders.iter() {
+                match order {
+                    model::solver_competition::Order::Colocated {
+                        id,
+                        sell_amount,
+                        buy_amount,
+                    } => {
+                        orders.insert(
+                            domain::OrderUid(id.0),
+                            competition::TradedAmounts {
+                                sell: (*sell_amount).into(),
+                                buy: (*buy_amount).into(),
+                            },
+                        );
+                    }
+                    model::solver_competition::Order::Legacy {
+                        id: _,
+                        executed_amount: _,
+                    } => return Err(error::Solution::NotFound),
```

Comment:
Maybe something related to the inconsistent state/unexpected results?

Reply:
This was already suggested by Felix to be not found. I think this should never be hit and I even expect this code to be removed once we save winning solutions to separate db table.